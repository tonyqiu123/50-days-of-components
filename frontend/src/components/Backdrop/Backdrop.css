.backdrop {
    pointer-events: none;
    opacity: 0;
    z-index: 1000;
    position: fixed;
    background-color: rgba(255, 255, 255, 0.808);
    backdrop-filter: blur(2px);
    inset: 0;
    height: 100vh;
    width: 100vw;
    transition: all 250ms cubic-bezier(0.25, 0.1, 0.25, 1);
}

.backdrop.darkMode {
    background-color: rgba(0, 0, 0, 0.596);
}

.showBackdrop {
    pointer-events: unset;
    opacity: 1;
}