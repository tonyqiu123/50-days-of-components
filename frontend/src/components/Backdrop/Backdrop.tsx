import React, { useEffect, useState } from 'react';
import './Backdrop.css'


interface BackdropProps {
    setShowBackdrop: React.Dispatch<React.SetStateAction<boolean>>;
    showBackdrop: boolean
    darkMode?: boolean
}

const Backdrop: React.FC<BackdropProps> = ({ setShowBackdrop, showBackdrop, darkMode = false }) => {

    return (
        <div onClick={() => setShowBackdrop(false)} className={`${darkMode && 'darkMode'} backdrop ${showBackdrop ? 'showBackdrop' : ''}`}>
        </div>
    );
};

export default Backdrop; 
