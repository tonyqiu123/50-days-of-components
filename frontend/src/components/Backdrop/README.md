# Day 20/100

July 27th / October 16th

# Backdrop
<a href="https://youtu.be/N4yV-UQnAFA" target="_blank">Watch live demo on youtube</a>

<a href="https://100daysofcomponents.netlify.app/Backdrop" target="_blank">Demo it yourself</a>

<a href="https://100daysofcomponents.netlify.app/Backdrop" target="_blank"><img src="https://cdn.discordapp.com/attachments/715319623637270638/1134247172527902850/image.png"/></a>  

# Description 

Day 20, one fifth complete of the journey! 
The "Backdrop" component serves to display a blurred, 'pointer-event-less' overlap on the web. Used in nearly all applications, and used in many components including modals, sheets, alerts, and dialogs.

# Installation 

To use the Backdrop component in your project, follow these steps:

1. Create a new folder called 'Backdrop' in your project's components directory.
2. Copy the `Backdrop.tsx` and `Backdrop.css` files into the newly created 'Backdrop' folder.

# Props 
### Backdrop
`showBackdrop` (required boolean): This property dictates whether the backdrop component should be displayed or not. If true, the backdrop is visible.

`setShowBackdrop` (required function): This is a function that sets the state of the showBackdrop property. It takes a boolean value as argument to toggle the visibility of the Backdrop component.

`darkMode` (optional boolean): This prop controls the theme mode of the backdrop. If set to true, the backdrop will display a dark theme. The default theme is light if this property is set to false or not provided.


# Example Usage
### page.tsx
```jsx
'use client'

import Button from '@/components/Button/Button';
import Backdrop from '@/components/Backdrop/Backdrop';

const BackdropDemo: React.FC = () => {

    const [isDarkMode, setIsDarkMode] = useState(false)
    const [showBackdrop, setShowBackdrop] = useState(false)

    return (
        <>
            <Button variant='primary' text={isDarkMode ? 'Untoggle dark mode' : 'Toggle dark mode'} handleClick={async () => setIsDarkMode(!isDarkMode)} />
            <Button handleClick={async () => setShowBackdrop(true)} variant='secondary' text='Open Backdrop' />
            <Backdrop darkMode={isDarkMode} showBackdrop={showBackdrop} setShowBackdrop={setShowBackdrop} />;
        </>
)}

export default BackdropDemo;
```

# Prerequisites
This component requires Node v14 or later to work correctly with NextJS framework version ^13
The following dependencies were used for this component:
1. "@types/node": "20.4.1",
2. "@types/react": "18.2.14",
3. "@types/react-dom": "18.2.6",
4. "next": "13.4.9",
5. "react": "18.2.0",
6. "react-dom": "18.2.0",
7. "typescript": "5.1.6"