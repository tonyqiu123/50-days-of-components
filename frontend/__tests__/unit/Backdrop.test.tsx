import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Backdrop from '@/components/Backdrop/Backdrop';

describe('Backdrop component', () => {
  it('should render correctly with default props', () => {
    const setShowBackdrop = jest.fn();
    const { container } = render(<Backdrop setShowBackdrop={setShowBackdrop} showBackdrop={true} />);
    
    expect(container).toMatchSnapshot();
  });

  it('should call setShowBackdrop when clicked', () => {
    const setShowBackdrop = jest.fn();
    const { container } = render(<Backdrop setShowBackdrop={setShowBackdrop} showBackdrop={true} />);
    
    const backdropElement = container.querySelector('.backdrop');
    if (backdropElement) {
      fireEvent.click(backdropElement);
      expect(setShowBackdrop).toHaveBeenCalledTimes(1);
      expect(setShowBackdrop).toHaveBeenCalledWith(false);
    }
  });

  it('should apply darkMode class when darkMode prop is true', () => {
    const setShowBackdrop = jest.fn();
    const { container } = render(<Backdrop setShowBackdrop={setShowBackdrop} showBackdrop={true} darkMode={true} />);
    
    const backdropElement = container.querySelector('.backdrop');
    expect(backdropElement).toHaveClass('darkMode');
  });

  it('should include additional classes passed via className prop', () => {
    const setShowBackdrop = jest.fn();
    const { container } = render(<Backdrop setShowBackdrop={setShowBackdrop} showBackdrop={true} className="custom-class" />);
    
    const backdropElement = container.querySelector('.backdrop');
    expect(backdropElement).toHaveClass('custom-class');
  });

  it('should have "showBackdrop" class when showBackdrop prop is true', () => {
    const setShowBackdrop = jest.fn();
    const { container } = render(<Backdrop setShowBackdrop={setShowBackdrop} showBackdrop={true} />);
    
    const backdropElement = container.querySelector('.backdrop');
    expect(backdropElement).toHaveClass('showBackdrop');
  });
});
