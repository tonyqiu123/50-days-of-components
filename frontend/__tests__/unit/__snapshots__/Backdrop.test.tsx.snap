// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop component should render correctly with default props 1`] = `
<div>
  <div
    class="  false backdrop showBackdrop"
  />
</div>
`;
